Mgvalleys: Add Dry Riverbeds

duane-r opened the pull request:
Lower water table where base humidity is low. Alter biome noises to
compensate for river humidity and altitude chill. Correct misuse of
surface_max_y in generateTerrain. Remove sand trails in the water at
river mouths. Make humidity calculations more readable.

Changes to the biome noises were based on repeated testing. However, the humidity scale of *0.8 fits well with the average river humidity increase of 1.0 - 1.5. These changes and increasing the humidity_dropoff seem to help the issue of jungle trees showing up more often around rivers since commit 80c7612e.

![sample](https://forum.minetest.net/download/file.php?mode=view&id=5023&sid=0ccd739855bc8a3db3c4a97dad0f00cf)

nerzhul commented on src/mapgen_valleys.cpp:
```diff
@@ -552,6 +564,11 @@ float MapgenValleys::terrainLevelAtPoint(s16 x, s16 z)
+	// Raising this reduces the rate of evaporation.
+	const float evaporation = 200.f;
+	// from the lua
+	const float humidity_dropoff = 4.f;
```
make this static

nerzhul commented on src/mapgen_valleys.cpp:
```diff
@@ -552,6 +564,11 @@ float MapgenValleys::terrainLevelAtPoint(s16 x, s16 z)
 int MapgenValleys::generateTerrain()
 {
+	// Raising this reduces the rate of evaporation.
+	const float evaporation = 200.f;
```
make this static

duane-r commented:
I just spotted a very odd issue this morning -- straight lines appearing in the landscape. They don't happen often, and they're harder to spot with decorations on.
![example](http://duanerobertson.com/img/c-demo-86.jpg)
The fix seems to be to put the offsets in the noises after generating, rather than changing the noise parameters, though I'm not sure why that would be. Anyway, I've made the change.

paramat commented:
Surely the change to noise parameters has to be done in the constructor before the noise objects are created? Changing them in 'calculateNoise' is too late?

est31 commented:
newer version of #3711

duane-r commented:
>Surely the change to noise parameters has to be done in the constructor before the noise objects are created? Changing them in 'calculateNoise' is too late?

I was changing the noise parameters in the constructor, but for some reason that produces artifacts. Changing the numbers in calculateNoise makes everything random again. /shrug

duane-r commented:
I changed the minimum evaporation from 0 to 0.08.

paramat commented:
:+1: Once i merge #3755 and you rebase on that.

paramat commented:
Yes this is better than changeing the parameters.

This PR will conflict with #3755 so please can you pull that PR into this? it has a simpler fix for river water by just moving the placement of river water to after the placement of lake water, so automatically eliminating river water below 'water_level':
```
				if (river && y == surface_y) {
					// river bottom
					vm->m_data[index_data] = n_sand;
				} else if (river && y <= surface_y) {
					// ground
					vm->m_data[index_data] = n_stone;
				} else if ((!river) && myround(fill * slope) >= y - surface_y) {
					// ground
					vm->m_data[index_data] = n_stone;
					heightmap[index_2d] = surface_max_y = y;
				} else if (y <= water_level) {
					// sea
					vm->m_data[index_data] = n_water;
				} else if (river && y < river_y) {
					// river
					vm->m_data[index_data] = n_river_water;
				} else {
					vm->m_data[index_data] = n_air;
				}
```
This is the standard method using the order ground-ocean-river-air.

paramat commented:
+1 once those 2 issues are attended to.

duane-r commented:
I've added Gael-de-Sailly's #3755, removed two unnecessary checks against the river variable, and reverted the changes to generateBiomes (except for using fill instead of sand). I changed the humidity offset slightly based on my impression of getting more desert (rather than hard numbers).

paramat commented:
:+1:

duane-r commented:
I found and corrected another bug in heightmap that I'd been looking for, and simplified the stone placement in generateTerrain. It seems to run noticeably faster now.

paramat commented:
:+1:

Gael-de-Sailly commented:
Another bug that only happens with your `dry_rivers` branch:
![bug](https://cloud.githubusercontent.com/assets/6905002/13202762/edbd94c2-d8a6-11e5-8269-3ea67ec2034c.png)
Where a river reaches the sea, jungle trees can grow in the river (with the biome parameters you linked [here](https://forum.minetest.net/viewtopic.php?p=207701#p207701) and `mg_valleys_river_size = 8`), only at y=0.

duane-r commented:
>Where a river reaches the sea, jungle trees can grow in the river

That was my intention. I wanted to get rid of the sand near river mouths where a swamp biome comes up, and place the appropriate soil, then let it act like any other part of the swamp. It bothered me that the rivers broke the swamp's look up. However, I still want to keep trees out of the river for most biomes -- placing sand is the easiest way to do that.

It would be nice to let just the jungle trees grow all along the river, but I don't see any neat way to do that.

duane-r commented:
I changed the altitude-based heat/humidity calculations to correct a problem Gael-de-Sailly found -- reductions were being reversed when the noise value was less than zero. I also corrected the stone placement, which was causing occasional river overflows.

paramat commented:
:+1:

duane-r commented:
I corrected problems with river over/underflows caused by rounding or truncating river_y and surface_y. The lua uses generic variables, so no integer conversions are involved. Using floats for these gives much better results.

paramat commented:
Good idea. Apart from my line comment :+1:

paramat commented:
:+1:

duane-r commented:
Corrected an issue where heightmap was still being set too high when slope * fill became negative.

I'm ambivalent about the approximation based on surface_y.
```
if (heightmap[index_2d] == -MAX_MAP_GENERATION_LIMIT)
        heightmap[index_2d] = myround(surface_y);
```
It's not being used for biomes and it produces confusing results in the lua voxelmanip functions, so my first impulse is to remove it.

Sorry about all the bugs.

paramat commented:
No problem, it's ongoing improvements so expected.

paramat commented:
I'll wait for this to settle for a couple of days, then merge.

duane-r commented:
I've made one more change, to quash that bad heightmap issue. It's now setting heightmap to values that are definitely out of range of the chunk when the actual terrain must be outside. This will avoid confusing lua mods as in https://forum.minetest.net/viewtopic.php?f=47&t=14111. They'll still get an approximate value, but they won't be tempted to use it for setting up terrain.

paramat commented:
Good it's now consistent with 'update heightmap' that other mapgens use, that returns -maxmapgenlimit if nothing is found between ymax and ymin.

paramat commented:
:+1:

paramat commented:
Approval from est31 http://irc.minetest.ru/minetest-dev/2016-02-27#i_4546393
Will merge soon.

paramat commented:
6969dd4224d0f309e006edf50b8eb1a188fc1254

Gael-de-Sailly commented:
:)
